### PR TITLE
MaterialIDs for solid models

### PR DESCRIPTION
**Pull request by endJunction**
Does not change the current behaviour: A single constitutive relation is used independent of MaterialIDs being present or not, if there is only one `<constitutive_relation>` section.
Adding an "id" attribute to it `<constitutive_relation id="42">` __and__ having multiple constitutive relations requires the MaterialIDs to be available and there must be a one-to-one correspondence.

Added a simple test gravity with two different materials having different E moduli.
Modified a CreepBGRa test, which was using a special case of the BGRa model (A = 0) to simulate the linear elastic case. Now there are two different materials.

### CODE REVIEW

File: MaterialLib/SolidModels/CreateConstitutiveRelation.cpp [hunk 1]
```diff
@@ -33,47 +33,83 @@ createConstitutiveRelation(
...
+        //! \ogs_file_param{material__solid__constitutive_relation}
+        config.getConfigSubtreeList("constitutive_relation");
+
+    std::map<int, std::unique_ptr<MechanicsBase<DisplacementDim>>>
+        constitutive_relations;
+
+    for (auto const& constitutive_relation_config :
+         constitutive_relation_configs)
+    {
+        int material_id =
```
Comment (endJunction) [inline MaterialLib/SolidModels/CreateConstitutiveRelation.cpp]: const

File: ProcessLib/HydroMechanics/CreateHydroMechanicsProcess.cpp [hunk 2]
```diff
@@ -101,9 +101,14 @@ std::unique_ptr<Process> createHydroMechanicsProcess(
...
-    auto material =
-        MaterialLib::Solids::createConstitutiveRelation<DisplacementDim>(
+    auto const material_ids =
+        mesh.getProperties().existsPropertyVector<int>("MaterialIDs")
+            ? mesh.getProperties().getPropertyVector<int>("MaterialIDs")
+            : nullptr;
+
+    // Constitutive relations.
+    auto solid_materials =
+        MaterialLib::Solids::createConstitutiveRelations<DisplacementDim>(
```
Comment (endJunction) [inline ProcessLib/HydroMechanics/CreateHydroMechanicsProcess.cpp]: One could collect `material_ids` and `solid_materials` in single class and add the above `select...` function to it...

File: ProcessLib/LIE/SmallDeformation/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h [hunk 3]
```diff
@@ -81,9 +81,12 @@ SmallDeformationLocalAssemblerMatrixNearFracture<ShapeFunction,
     _ip_data.reserve(n_integration_points);
     _secondary_data.N.resize(n_integration_points);
 
+    auto const material_id = _process_data.material_ids == nullptr
```
Comment (endJunction) [inline ProcessLib/LIE/SmallDeformation/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h]: Use select function here.

File: ProcessLib/RichardsMechanics/CreateRichardsMechanicsProcess.cpp [hunk 4]
```diff
@@ -174,14 +174,15 @@ std::unique_ptr<Process> createRichardsMechanicsProcess(
         //! \ogs_file_param{prj__processes__process__RICHARDS_MECHANICS__material_property}
         config.getConfigSubtree("material_property");
 
-    boost::optional<MeshLib::PropertyVector<int> const&> material_ids;
-    if (mesh.getProperties().existsPropertyVector<int>("MaterialIDs"))
+    auto const material_ids =
+        mesh.getProperties().existsPropertyVector<int>("MaterialIDs")
+            ? mesh.getProperties().getPropertyVector<int>("MaterialIDs")
+            : nullptr;
```
Comment (endJunction) [inline ProcessLib/RichardsMechanics/CreateRichardsMechanicsProcess.cpp]: use select here too.

File: ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h [hunk 5]
```diff
@@ -54,10 +54,12 @@ RichardsMechanicsLocalAssembler<ShapeFunctionDisplacement,
                           IntegrationMethod, DisplacementDim>(
             e, is_axially_symmetric, _integration_method);
 
+    auto const material_id = _process_data.material_ids == nullptr
```
Comment (endJunction) [inline ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h]: use select here too.

File: ProcessLib/HydroMechanics/HydroMechanicsFEM-impl.h [hunk 6]
```diff
@@ -53,10 +54,12 @@ HydroMechanicsLocalAssembler<ShapeFunctionDisplacement, ShapeFunctionPressure,
                           IntegrationMethod, DisplacementDim>(
             e, is_axially_symmetric, _integration_method);
 
+    auto& solid_material = MaterialLib::Solids::selectSolidConstitutiveRelation(
```
Comment (wenqing) [inline ProcessLib/HydroMechanics/HydroMechanicsFEM-impl.h]: const?

File: ProcessLib/HydroMechanics/HydroMechanicsProcessData.h [hunk 7]
```diff
@@ -82,10 +72,14 @@ struct HydroMechanicsProcessData
     //! Assignments are not needed.
     void operator=(HydroMechanicsProcessData&&) = delete;
 
+    MeshLib::PropertyVector<int> const* material_ids;
```
Comment (wenqing) [inline ProcessLib/HydroMechanics/HydroMechanicsProcessData.h]: Can the type be reference?

File: ProcessLib/LIE/SmallDeformation/CreateSmallDeformationProcess.cpp [hunk 8]
```diff
@@ -98,9 +98,14 @@ std::unique_ptr<Process> createSmallDeformationProcess(
         process_variables;
     process_variables.push_back(std::move(per_process_variables));
 
+    auto const material_ids =
```
Comment (wenqing) [inline ProcessLib/LIE/SmallDeformation/CreateSmallDeformationProcess.cpp]: Can it be a member of mesh class? It appears in several functions.

Comment (wenqing) [reply ProcessLib/LIE/SmallDeformation/CreateSmallDeformationProcess.cpp]: I mean to add a member to Mesh as
```
PropertyVector<int> Mesh::getMaterialIDs() const
{
return _properties.existsPropertyVector<int>("MaterialIDs")
            ? _properties.getPropertyVector<int>("MaterialIDs")
            : nullptr;
}
```

File: ProcessLib/HydroMechanics/HydroMechanicsFEM-impl.h (hunk 6 above)
Comment (endJunction) [reply ProcessLib/HydroMechanics/HydroMechanicsFEM-impl.h]: This is actually possible. Did mark the integrateStress and create*State functions as const, then this one and in the IPData too...

File: ProcessLib/HydroMechanics/HydroMechanicsProcessData.h (hunk 7 above)
Comment (endJunction) [reply ProcessLib/HydroMechanics/HydroMechanicsProcessData.h]: No, not with current implementation. It is allowed not to have the 'material_ids'.

File: ProcessLib/LIE/SmallDeformation/CreateSmallDeformationProcess.cpp (hunk 8 above)
Comment (endJunction) [reply ProcessLib/LIE/SmallDeformation/CreateSmallDeformationProcess.cpp]: What do other think if we add a free function in MeshLib namespace?
```c++
PropertyVector<int>* materialIDs(Mesh const&);
```

Comment (wenqing) [reply ProcessLib/LIE/SmallDeformation/CreateSmallDeformationProcess.cpp]: Sounds good.

File: ProcessLib/HydroMechanics/CreateHydroMechanicsProcess.cpp [hunk 9]
```diff
@@ -187,15 +192,11 @@ std::unique_ptr<Process> createHydroMechanicsProcess(
...
-        fluid_density,
-        biot_coefficient,
-        porosity,
-        solid_density,
-        specific_body_force,
+        material_ids,           std::move(solid_materials),
+        intrinsic_permeability, specific_storage,
+        fluid_viscosity,        fluid_density,
+        biot_coefficient,       porosity,
+        solid_density,          specific_body_force,
```
Comment (TomFischer) [inline ProcessLib/HydroMechanics/CreateHydroMechanicsProcess.cpp]: Is the formatting okay?

Comment (endJunction) [reply ProcessLib/HydroMechanics/CreateHydroMechanicsProcess.cpp]: Looks good to clang-format...

**Pull request merged**